The Following Code works on Google Colab and uploading on Tweets.csv file

1.Import the libraries, load dataset, print shape of data, data description.



In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
data = pd.read_csv('Tweets.csv')
data.head()

tweet_id  ...               user_timezone
0  570306133677760513  ...  Eastern Time (US & Canada)
1  570301130888122368  ...  Pacific Time (US & Canada)
2  570301083672813571  ...  Central Time (US & Canada)
3  570301031407624196  ...  Pacific Time (US & Canada)
4  570300817074462722  ...  Pacific Time (US & Canada)

[5 rows x 15 columns]

In [3]:
data.shape

(14640, 15)

In [4]:
data.describe()

tweet_id  ...  retweet_count
count  1.464000e+04  ...   14640.000000
mean   5.692184e+17  ...       0.082650
std    7.791112e+14  ...       0.745778
min    5.675883e+17  ...       0.000000
25%    5.685592e+17  ...       0.000000
50%    5.694779e+17  ...       0.000000
75%    5.698905e+17  ...       0.000000
max    5.703106e+17  ...      44.000000

[8 rows x 4 columns]

# **2.Understand of data-columns:**

In [5]:
data.columns

Index(['tweet_id', 'airline_sentiment', 'airline_sentiment_confidence',
       'negativereason', 'negativereason_confidence', 'airline',
       'airline_sentiment_gold', 'name', 'negativereason_gold',
       'retweet_count', 'text', 'tweet_coord', 'tweet_created',
       'tweet_location', 'user_timezone'],
      dtype='object')

a.Drop all other columns except “text” and “airline_sentiment”.

In [6]:
data.drop(data.columns.difference(['text','airline_sentiment']), 1, inplace=True)

In [7]:
data.shape

(14640, 2)

In [8]:
data.head()

airline_sentiment                                               text
0           neutral                @VirginAmerica What @dhepburn said.
1          positive  @VirginAmerica plus you've added commercials t...
2           neutral  @VirginAmerica I didn't today... Must mean I n...
3          negative  @VirginAmerica it's really aggressive to blast...
4          negative  @VirginAmerica and it's a really big bad thing...

# **3. Text pre-processing: Data preparation.**

In [9]:
# imports
from bs4 import BeautifulSoup
# function to remove HTML tags
def remove_html_tags(text):
    return BeautifulSoup(text, 'html.parser').get_text()

In [10]:
# Rwmove html tags
for i in range(len(data['text'])):
  data['text'][i] = remove_html_tags(data['text'][i])

In [11]:
#conda install -c anaconda nltk
import nltk
nltk.download()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> d

Download which package (l=list; x=cancel)?
  Identifier> l
Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [

True

In [12]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [13]:
from nltk.tokenize import word_tokenize
token = []
for i in range(len(data['text'])):
  token.append(word_tokenize(data['text'][i]))

In [14]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [15]:
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
for i in range(len(data['text'])):
  filtered_sentence = [w for w in token[i] if not w in stop_words] 
  token[i] = filtered_sentence

In [16]:
import string
# remove all tokens that are not alphabetic
words = []
for i in range(len(data['text'])):
  words.append([word for word in token[i] if word.isalpha()])
  for j in range(len(words[i])):
    words[i][j] = words[i][j].lower()

In [17]:
from nltk.stem import LancasterStemmer,PorterStemmer,SnowballStemmer
Pst = PorterStemmer()

In [18]:
for i in range(len(words)):
  for j in range(len(words[i])):
    words[i][j] = Pst.stem(words[i][j])

In [20]:
data2 = data
for i in range(len(words)):
  a = ""
  for j in range(len(words[i])):
     a=a + words[i][j] + " "
  data2['text'][i] = a

In [21]:
data2.head() #After Preprocessing

airline_sentiment                                               text
0           neutral                  virginamerica what dhepburn said 
1          positive        virginamerica plu ad commerci experi tacki 
2           neutral  virginamerica i today must mean i need take an...
3          negative  virginamerica realli aggress blast obnoxi ente...
4          negative                virginamerica realli big bad thing

In [22]:
X = data2['text']
Y = data2['airline_sentiment']

4.    Vectorization
5.    Fit and evaluate model using both type of vectorization.

In [23]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(lowercase=True, stop_words='english', min_df=2)
X_cv = cv.fit_transform(X.values.astype('U'))
X_cv.shape
X_names = cv.get_feature_names()
X_cv_df = pd.DataFrame(X_cv.toarray(),columns = X_names)
X_cv_df.describe()

aa          aadv  ...          zoom        zurich
count  14640.000000  14640.000000  ...  14640.000000  14640.000000
mean       0.013934      0.000137  ...      0.000137      0.000273
std        0.125660      0.011688  ...      0.011688      0.016528
min        0.000000      0.000000  ...      0.000000      0.000000
25%        0.000000      0.000000  ...      0.000000      0.000000
50%        0.000000      0.000000  ...      0.000000      0.000000
75%        0.000000      0.000000  ...      0.000000      0.000000
max        3.000000      1.000000  ...      1.000000      1.000000

[8 rows x 4132 columns]

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
xT, xt, yT, yt = train_test_split(X_cv_df, Y, test_size=0.30, random_state=11)
model = LogisticRegression()
model.fit(xT, yT)
model.predict(xt)
pd.DataFrame(model.predict(xt))
model.score(xT,yT)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8966627634660421

In [25]:
res = pd.DataFrame(model.predict(xt))
res

0
0     positive
1     negative
2     negative
3     negative
4     positive
...        ...
4387  negative
4388  negative
4389  negative
4390  negative
4391  negative

[4392 rows x 1 columns]

In [26]:
pd.DataFrame(model.predict_proba(xt))

0         1         2
0     0.012397  0.012299  0.975304
1     0.990027  0.003772  0.006201
2     0.906792  0.066127  0.027081
3     0.998449  0.001496  0.000055
4     0.028212  0.026576  0.945212
...        ...       ...       ...
4387  0.597264  0.162959  0.239777
4388  0.995104  0.004345  0.000551
4389  0.637274  0.021400  0.341326
4390  0.590242  0.244241  0.165517
4391  0.996735  0.000122  0.003143

[4392 rows x 3 columns]

In [27]:
model.score(xt, yt)

0.7800546448087432

In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer
tv = TfidfVectorizer(lowercase=True, stop_words='english',min_df=2)
tv_1 = tv.fit_transform(X)
tv_1.shape

(14640, 4132)

In [29]:
tv_1_names = tv.get_feature_names()
tv_1_df = pd.DataFrame(tv_1.toarray(),columns= tv_1_names)
tv_1_df.describe()

aa          aadv  ...          zoom        zurich
count  14640.000000  14640.000000  ...  14640.000000  14640.000000
mean       0.004016      0.000055  ...      0.000054      0.000110
std        0.036735      0.004724  ...      0.004600      0.006662
min        0.000000      0.000000  ...      0.000000      0.000000
25%        0.000000      0.000000  ...      0.000000      0.000000
50%        0.000000      0.000000  ...      0.000000      0.000000
75%        0.000000      0.000000  ...      0.000000      0.000000
max        0.916388      0.404156  ...      0.409715      0.460871

[8 rows x 4132 columns]

In [30]:
x_train,x_test,y_train,y_test = train_test_split(tv_1_df,Y,test_size = 0.30, random_state = 11)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
model.score(x_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8534348165495707

In [31]:
model.score(x_test, y_test)

0.788023679417122

In [32]:
Algo = []
train = []
test = []
recall = []
roc = []

In [33]:
from sklearn.metrics import recall_score,precision_score,roc_auc_score
Algo.append('Logistic Regression')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test),average = 'micro')
#roc.append(roc_auc_score(ypredict,y_test))


In [34]:
from sklearn.tree import DecisionTreeClassifier

In [35]:
model_decisiontree = DecisionTreeClassifier()
model_decisiontree.fit(x_train,y_train)
ypredict_decisiontree = model_decisiontree.predict(x_test)
print(model_decisiontree.tree_.max_depth)
#model_decisiontree.score(x_train,y_train) Overfitting of training set
#model_decisiontree.score(x_test,y_test)

206


In [36]:
model_pruned = DecisionTreeClassifier(criterion='entropy',max_depth=3,random_state=5)

In [37]:
model_pruned.fit(x_train,y_train)
ypredict = model_pruned.predict(x_test)

In [38]:
Algo.append('Decision Tree using Entropy')
train.append(model_pruned.score(x_train,y_train))
test.append(model_pruned.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test))
#roc.append(roc_auc_score(ypredict,y_test))

In [39]:
model = DecisionTreeClassifier(criterion='gini',max_depth=3,random_state=5)
#model = DecisionTreemodel()
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Decision Tree using Gini')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test))
#roc.append(roc_auc_score(ypredict,y_test))

In [40]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators = 100)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Random Forest')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))

Following Classifier have low accuracy and are time consuming hence commented

In [41]:
#from sklearn.ensemble import BaggingClassifier
#model = BaggingClassifier(n_estimators=100,max_samples=0.7,bootstrap=True,oob_score=True,random_state=11)
#model.fit(x_train,y_train)
#ypredict = model.predict(x_test)
#Algo.append('Bagging model')
#train.append(model.score(x_train,y_train))
#test.append(model.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test))
#roc.append(roc_auc_score(ypredict,y_test))

In [42]:
#from sklearn.ensemble import AdaBoostClassifier
#model = AdaBoostClassifier(n_estimators=200,learning_rate=0.1,random_state=11)
#model.fit(x_train,y_train)
#ypredict = model.predict(x_test)
#Algo.append('AdaBoost model')
#train.append(model.score(x_train,y_train))
#test.append(model.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test))
#roc.append(roc_auc_score(ypredict,y_test))

In [43]:
#from sklearn.ensemble import GradientBoostingClassifier
#model = GradientBoostingClassifier(n_estimators=200,random_state=11)
#model.fit(x_train,y_train)
#ypredict = model.predict(x_test)
#Algo.append('Gradient Boosting model')
#train.append(model.score(x_train,y_train))
#test.append(model.score(x_test,y_test))
#recall.append(recall_score(ypredict,y_test))
#roc.append(roc_auc_score(ypredict,y_test))

In [44]:
results = pd.DataFrame()
results['Model'] = Algo
results['Training Accuracy'] = train
results['Testing Accuracy'] = test
#results['Recall'] = recall
#results['ROC AUC Score'] = roc
results = results.set_index('Model')
results

Training Accuracy  Testing Accuracy
Model                                                           
Logistic Regression                   0.853435          0.788024
Decision Tree using Entropy           0.678962          0.681922
Decision Tree using Gini              0.683743          0.686931
Random Forest                         0.993657          0.757058

6. The preprocessing of the text data is an essential step as it makes the raw text ready for mining, i.e., it becomes easier to extract information from the text and apply machine learning algorithms to it. If we skip this step then there is a higher chance that you are working with noisy and inconsistent data.Vectorization is to Convert the data into Numerical Matrix in which ML Algorithms can easily be applied

In this Case we find Logistic Regression to be most effective with 78.8% Accuracy

In [45]:
for i in range(len(Y)):
  if Y[i] == 'positive':
    Y[i] = 0.0
  if Y[i] == 'negative':
    Y[i] = 1.0
  if Y[i] == 'neutral':
    Y[i] = 2.0

# **7.    If applicable apply any Classification based Machine learning/Deep Learning framework to have better understanding of the Use Case.**

In [46]:
clf = RandomForestClassifier(
    n_estimators=50,
    criterion='gini',
    max_depth=5,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features='auto',
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    min_impurity_split=None,
    bootstrap=True,
    oob_score=False,
    n_jobs=-1,
    random_state=0,
    verbose=0,
    warm_start=False,
    class_weight='balanced'
)

In [47]:
#from sklearn.model_selection import StratifiedKFold
#cv = StratifiedKFold(n_splits=5, random_state=123, shuffle=True)
#results = pd.DataFrame(columns=['training_score', 'test_score'])
#fprs, tprs, scores = [], [], []
    
#for (train1, test1), i in zip(cv.split(X_cv_df, Y), range(5)):
#    clf.fit(X_cv_df.iloc[train1], Y.iloc[train1])
#    train.append(clf.score(x_train,y_train))
#    test.append(clf.score(x_test,y_test))
#    Algo.append('RFCV')
#train.append(model.score(x_train,y_train))

In [48]:
from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import mean_squared_error, r2_score
ts_split= TimeSeriesSplit(n_splits=5)
for train_index, test_index in ts_split.split(X_cv_df):
        X_train, X_test = X_cv_df[:len(train_index)], X_cv_df[len(train_index): (len(train_index)+len(test_index))]
        y_train, y_test = Y[:len(train_index)].values.ravel(), Y[len(train_index): (len(train_index)+len(test_index))].values.ravel()

In [49]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12200 entries, 0 to 12199
Columns: 4132 entries, aa to zurich
dtypes: int64(4132)
memory usage: 384.6 MB


In [50]:
X_train =np.array(X_train)
X_test =np.array(X_test)

X_tr_t = X_train.reshape(X_train.shape[0], 1, X_train.shape[1])
X_tst_t = X_test.reshape(X_test.shape[0], 1, X_test.shape[1])

In [51]:
for i in range(len(y_train)):
  if y_train[i] == 'positive':
    y_train[i] = 0.0
  if y_train[i] == 'negative':
    y_train[i] = 1.0
  if y_train[i] == 'neutral':
    y_train[i] = 2.0
for i in range(len(y_test)):
  if y_test[i] == 'positive':
    y_test[i] = 0.0
  if y_test[i] == 'negative':
    y_test[i] = 1.0
  if y_test[i] == 'neutral':
    y_test[i] = 2.0
y_train

array([2.0, 0.0, 2.0, ..., 0.0, 1.0, 0.0], dtype=object)

In [52]:
X_train = np.asarray(X_train).astype(np.float32)
y_train = np.asarray(y_train).astype(np.float32)
y_test = np.asarray(y_test).astype(np.float32)
X_tr_t = np.asarray(X_tr_t).astype(np.float32)
X_tst_t = np.asarray(X_tst_t).astype(np.float32)

In [53]:
#import tensorflow as tf
#from tensorflow import keras
from sklearn.metrics import mean_squared_error, r2_score
from keras.layers import LSTM,Dense,Dropout
from keras.models import Sequential
from keras.layers import Dense
import keras.backend as K
from keras.callbacks import EarlyStopping
from keras.optimizers import Adam
from keras.models import load_model
from keras.layers import LSTM
K.clear_session()
model_lstm = Sequential()
model_lstm.add(LSTM(16, input_shape=(1, X_train.shape[1]), activation='relu', return_sequences=False))
model_lstm.add(Dense(1))
model_lstm.compile(loss='mean_squared_error', optimizer='adam')
early_stop = EarlyStopping(monitor='loss', patience=5, verbose=1)
history_model_lstm = model_lstm.fit(X_tr_t, y_train, epochs=25, batch_size=8, verbose=1, shuffle=False, callbacks=[early_stop])

Epoch 1/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.4006
Epoch 2/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.2578
Epoch 3/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.2221
Epoch 4/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1941
Epoch 5/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1701
Epoch 6/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1496
Epoch 7/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1333
Epoch 8/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1200
Epoch 9/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1093
Epoch 10/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.1005
Epoch 11/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.0926
Epoch 12/25
1525/1525 [==============================] - 5s 3ms/step - loss: 0.0868
E

In [55]:
y_pred_test_lstm = model_lstm.predict(X_tst_t)
y_train_pred_lstm = model_lstm.predict(X_tr_t)
print("The R2 score on the Train set is:\t{:0.3f}".format(r2_score(y_train, y_train_pred_lstm)))
r2_train = r2_score(y_train, y_train_pred_lstm)

The R2 score on the Train set is:	0.883
